### PR TITLE
Make attrs in RuleInfo more precise

### LINKED ISSUE #30: More precise type for RuleInfo regarding attrs
Currently, `attrs` in `RuleInfo` is a list of text-keyed tuples. In [reality](https://github.com/tweag/gazelle_cabal/blob/4d3cba80588c054bd97ec8fe2a88fd75fa6f8c21/cabalscan/src/CabalScan/RuleGenerator.hs#L189), this is always a list containing at most 4 things, with known keys. I think it would be an improvement to instead split `attrs` out to these 4 things.

Pros:
* more accurately models reality (?)
* it makes the code clearer
  * you have to scour less to find out what `attrs` could be
  * more descriptive names
  * easier testing (can directly access the field you want to check)
  * immediately obvious we don't handle all possible attributes
  
Cons:
* more code
* more code to add a new attribute

### PR DESCRIPTION
**Pull request by googleson78**
Closes #30

### CODE REVIEW

**Comment by googleson78** [comment]
The docs on `RuleInfo` are out of date, I should at least
* fix the example to use the new attributes
* remove the `ghcopts` from the example - it's no longer in `RuleInfo`

**Comment by kczulko** [comment]
I'm wondering howto address the feature like this: https://github.com/tweag/gazelle_cabal/pull/25 when we get this change merged to master... I'm thinking about the scenario when we'd like to add some attribute that is specific only to e.g. libraries or benchmarks. Current API, abstracting from types, gives some kind of flexibility on the corresponding `generate(Library|Benchmark|Test|Binary)Rule` function.
WDYT?

**Comment by googleson78** [comment]
Seems like you can add another field to `RuleInfo` regarding main files, right? I agree that it's more flexible, but in general I'd prefer not to have string keys when you know all the keys you're going to be using beforehand and they're not that much. I think it's also easier for a newcommer to the project to trace what the possible attributes are.

**Comment by kczulko** [comment]
Ah yes... but I was thinking about some other solution than adding any arbitrary field to `RuleInfo` and e.g. marking it as `Maybe` (e.g. main_file being ignored in `generateLibraryRule` but since things are lazy it doesn't matter so much?). Typeclasses? Maybe yes but IMHO it complicates than improves the situation. However, don't treat my voice as being disruptive for this improvement. I'm just loudly thinking and judging pros & cons :)

**Comment by googleson78** [comment]
If we want field specific to certain rules I'd do something (but with better names or something) like
```haskell
data MainSpecificFields = MainSpecificFields
  { main_is :: Text
  , ...others...
  }
data TestSpecificFields = TestSpecificFields
  { main_is :: Text
  , ...others...
  }
  
data RuleSpecific = MainSpecific MainSpecificFields | TestSpecific TestSpecificFields

data RuleInfo = RuleInfo
  { ...all the common stuff...
  , ruleSpecificAttrs :: RuleSpecific
  , ...other common stuff...
  }
```
Maybe even without the "intermediate" types, but that always brings the danger of partial field access (unless we turn on `NoFieldSelectors` at some point).

I'm not certain whether any of the current fields are specific to only one kind of rule/component, although they don't seem to be :thinking:

**Comment by kczulko** [comment]
> I'm not certain whether any of the current fields are specific to only one kind of rule/component, although they don't seem to be thinking

From the cabal's perspective, I think most of the fields we use are taken from [`BuildInfo`](https://hackage.haskell.org/package/Cabal-3.6.3.0/docs/Distribution-Types-BuildInfo.html#t:BuildInfo)

But sometimes build info is being aggregated by some other structures like [`Benchmark`](https://hackage.haskell.org/package/Cabal-3.6.3.0/docs/Distribution-Types-Benchmark.html#v:benchmarkInterface) which allows you to obtain e.g. `main-is` from e.g. [`benchmarkInterface`](https://hackage.haskell.org/package/Cabal-3.6.3.0/docs/Distribution-Types-BenchmarkInterface.html#t:BenchmarkInterface)

Similar situation applies for binary components.

I think that you're right at this point that there might not be much of those strange fields (like `main-is`). Maybe `main-is` is the only one here since vast majority (almost all except problematic `main-is`) are taken from `BuilInfo` structure.

File: cabalscan/src/CabalScan/Rules.hs [hunk 1]
```diff
@@ -16,27 +17,33 @@ import Data.Text (Text)
...
 --
--- stands for the rule instantiation
+-- stands for part of the rule instantiation
 --
 -- > haskell_library(
 -- >   name = 'foo',
 -- >   srcs = ['src/Main.hs'],
--- >   deps = ["@stackage//:protolude", "@libsodium//:libsodium"]
--- >   tools = ["@stackage-exe//tasty-discover"]
+-- >   hidden_modules = ["PackageA.Other.B", PackageA.Other.D"],
```
Comment (facundominguez) [inline cabalscan/src/CabalScan/Rules.hs:32]:
```suggestion
-- >   ghcopts = "-DFOO=1" 
-- >   hidden_modules = ["PackageA.Other.B", PackageA.Other.D"],
```

File: cabalscan/src/CabalScan/Rules.hs [hunk 2]
```diff
@@ -16,27 +17,33 @@ import Data.Text (Text)
 -- >   , deps = ["protolude"]
 -- >   , extraLibraries = ["libsodium"]
 -- >   , tools = [ToolName "tasty-discover" "tasty-discover"]
--- >   , attrs = [ ("srcs", ["src/Main.hs"])
--- >             , ("ghcopts", "-DFOO=1")
--- >             ]
+-- >   , version = "0.1.0.0"]
+-- >   , hidden_modules = Just ("PackageA.Other.B" :| ["PackageA.Other.D"])
+-- >   , dataAttr = Nothing
+-- >   , srcs = ["src/Main.hs"]
```
Comment (facundominguez) [inline cabalscan/src/CabalScan/Rules.hs:23]:
```suggestion
-- >   , ghcopts = ["-DFOO=1"]
```

Comment (facundominguez) [reply cabalscan/src/CabalScan/Rules.hs:23]: hm, ghcopts go elsewhere. This example is not up to date. We can either point that fields are missing or try to keep it up to date.

Comment (googleson78) [reply cabalscan/src/CabalScan/Rules.hs:23]: How do you want to bring it up to date? It would require also mentioning `ImportData`.

**Comment by googleson78** [comment]
Rebased on `main`, incorporating the `main-is` changes.

File: cabalscan/src/CabalScan/Rules.hs [hunk 3]
```diff
@@ -16,27 +17,34 @@ import Data.Text (Text)
 -- >   , deps = ["protolude"]
 -- >   , extraLibraries = ["libsodium"]
 -- >   , tools = [ToolName "tasty-discover" "tasty-discover"]
```
Comment (facundominguez) [inline cabalscan/src/CabalScan/Rules.hs:19]:
```suggestion
-- >   , importData = ImportData
-- >     { deps = ["protolude"]
-- >     , extraLibraries = ["libsodium"]
-- >     , ghcOpts = ["-Werror", "-Wall"]
-- >     , tools = [ToolName "tasty-discover" "tasty-discover"]
-- >     }
```

File: cabalscan/src/CabalScan/Rules.hs [hunk 4]
```diff
@@ -16,27 +17,34 @@ import Data.Text (Text)
...
--- >   , attrs = [ ("srcs", ["src/Main.hs"])
--- >             , ("ghcopts", "-DFOO=1")
--- >             ]
+-- >   , version = "0.1.0.0"]
+-- >   , hidden_modules = Just ("PackageA.Other.B" :| ["PackageA.Other.D"])
+-- >   , dataAttr = Nothing
```
Comment (facundominguez) [inline cabalscan/src/CabalScan/Rules.hs:22]:
```suggestion
-- >   , dataAttr = Nothing
-- >   , main_file = Nothing
```

**Comment by facundominguez** [comment]
Thanks @googleson78!

**Pull request merged**